ring.go

```diff
@@ -213,32 +215,54 @@ type ringShards struct {
 
 	mu       sync.RWMutex
 	hash     ConsistentHash
-	shards   map[string]*ringShard // read only
-	list     []*ringShard          // read only
+	shards   map[string]*ringShard // read only, updated by SetAddrs
+	list     []*ringShard          // read only, updated by SetAddrs
 	numShard int
 	closed   bool
 }
 
 func newRingShards(opt *RingOptions) *ringShards {
-	shards := make(map[string]*ringShard, len(opt.Addrs))
-	list := make([]*ringShard, 0, len(shards))
+	c := &ringShards{
+		opt: opt,
+	}
+	c.SetAddrs(opt.Addrs)
 
-	for name, addr := range opt.Addrs {
-		shard := newRingShard(opt, name, addr)
-		shards[name] = shard
+	return c
+}
 
-		list = append(list, shard)
+func (c *ringShards) SetAddrs(addrs map[string]string) {
+	newRingshards := make(map[string]*ringShard)
+	if c.closed {
+		return
 	}
 
-	c := &ringShards{
-		opt: opt,
+	for k, shard := range c.shards {
+		if addr, ok := addrs[k]; ok && shard.addr == addr {
+			newRingshards[k] = shard
+		} else {
+			defer shard.Client.Close()
+		}
+	}
 
-		shards: shards,
-		list:   list,
+	for k, addr := range addrs {
+		if shard, ok := c.shards[k]; !ok || shard.addr != addr {
```

Comment:
Should not you close the shard client when `shard.addr != addr`?

Reply:
Every existing shard with the name in the `addrs` map but different address should be closed by the `defer` in the previous loop.

Reply:
Every existing shard with the name in the `addrs` map but different address should be closed by the `defer` in the previous loop.